Define `CookieStore` as trait

pfernie opened the pull request:
This PR aims to address #1104 by replacing the newtype `reqwest::cookie::CookieStore` with `trait CookieStore` to define a minimal interface for `reqwest` to store and retrieve cookies.

The [cookie-storage branch of the `cookie_store` crate](https://github.com/pfernie/cookie_store/tree/cookie-storage) contains such an implementation.

Not addressed is providing a default implementation of a `CookieStore` within `reqwest`. Since the trait now reduces `reqwest`s external dependencies (`cookie_store` can be removed as a dependency for the `cookies` features), it may make sense to provide that as an alternative feature in `reqwest`. A new feature `default_cookie_store` could re-introduce the `cookie_store` dependency and provide it as the default `CookieStore` implementation.

* This PR still has a breaking API change in the `ClientBuilder`, as the `ClientBuilder::cookie_store` signature changed from accepting a `bool` to a `Option<dyn CookieStore>`. An alternative method (`set_cookie_store`) could be provided, but that would necessitate both multiple configuration options, and providing a default implementation unless the `bool` fn became gated behind the `default_cookie_store` feature as well.
* This PR would allow for the use case of read-only `CookieStore`. Trivially, a user could depend on `cookie_store::CookieStore`, wrap it in a newtype, and impl `reqwest::cookie::CookieStore` on the newtype implementing `store_cookie_headers` as a no-op.

---

Naming: in writing this description, it is somewhat confusing that the `cookie_store::CookieStore` implementation overlaps with the proposed trait `reqwest::cookie::CookieStore`. Documentation may be less confusing for the trait to retain the name `reqwest::cookie::CookieStorage`.

seanmonstar commented on src/cookie.rs:
```diff
@@ -90,14 +95,17 @@ pub(crate) fn extract_response_cookies<'a>(
-    }
+use std::any::Any;
+/// Actions for a persistent cookie store providing session supprt.
+pub trait CookieStore: fmt::Debug + Send + Sync {
```
Here's my suggestions around the actual trait:

- I'd remove the downcasting stuff. Instead of asking the `Client` for a reference, I would suggest that implementors could provide a type that has an `Arc` of the data or something, and they check their own clone from time to time.
- We could move the lock stuff to the implementation, by making the store method also take `&self`. I say this just since if people already have an `Arc<Mutex>`, it seems wasteful to then wrap *that* in an `RwLock`.
- Method names `set_cookies` and `cookies()`?
- I wonder about using `Vec`s of strings, versus slices of `HeaderValue`s... I haven't thought this one through too far.
- Probably remove the `Debug` bounds, at least.

seanmonstar reviewed (commented):
Thanks for the quick start! 💪 

Another question is around providing a "default" store. There is currently the `cookies` cargo feature to enable, so people who don't need it don't include the dependencies. With it being a trait also, we could perhaps include the trait always, and change the feature to include a default store. Or we could do as you suggest, splitting into 2 features...

pfernie commented:
Latest update addresses some points:
* Rename of method names (`set_cookies()`, `cookies()`)
* Remove the downcast / `Any` implementation; rather, change `cookie_store: Option<Arc<Mutex<dyn CookieStore>>>`. As you mention, we could remove the mention of `Mutex` by making `set_cookies(&self)` instead of `&mut self`. Being explicit seems a little more obvious to me here.

Not addressed:
* A default store - the `cookies` feature also enables `time` and the `cookie-rs` crates, which are still needed for `Response::cookies()` functionality. I think it would be worthwhile making a separate `default_cookie_store` feature to allow users to avoid the `cookie_store` dependency when not needed.
* Passing `&str/String` instead of `HeaderValue` in the trait fns - I was shooting for a lowest common denominator in this interface, but since it is `reqwest` exposing the trailt anyway, I would be fine making this change.

With this implementation, interaction with the store is straightforward:
```rust
    let cookie_store = cookie_store::CookieStore::default(); // or de-serialize from disk

    let cookie_store = Arc::new(Mutex::new(cookie_store));

    let client = reqwest::ClientBuilder::new()
        .cookie_store(Some(cookie_store.clone()))
        .build()?;

    {
        println!("fresh store, should be empty:");
        let cookie_store = cookie_store.lock().unwrap();
        assert_eq!(
            cookie_store.iter_any().count(),
            0,
            "fresh store should not have any cookies"
        );
        dbg!(&cookie_store);
    }

    client.get("https://google.com").send().await?;

    {
        let mut cookie_store = cookie_store.lock().unwrap();
        println!("just made a request, should have a cookie:");
        assert!(
            cookie_store.iter_any().count() != 0,
            "just made a request to google.com, should have some yummy cookies!"
        );
        dbg!(cookie_store
            .get("google.com", "/", "1P_JAR")
            .map(|c| c.to_string()));
        cookie_store.clear();
        println!("cleared store, should be empty:");
        assert_eq!(
            cookie_store.iter_any().count(),
            0,
            "cleared store should not have any cookies"
        );
        dbg!(&cookie_store);
    }
```
Producing:
```
fresh store, should be empty:
[src/main.rs:23] &cookie_store = CookieStore {
    cookies: {},
    public_suffix_list: None,
}
just made a request, should have a cookie:
[src/main.rs:35] cookie_store.get("google.com", "/", "1P_JAR").map(|c| c.to_string()) = Some(
    "1P_JAR=2020-12-10-15; Secure",
)
cleared store, should be empty:
[src/main.rs:45] &cookie_store = CookieStore {
    cookies: {},
    public_suffix_list: None,
}
```

seanmonstar commented:
> As you mention, we could remove the mention of Mutex by making set_cookies(&self) instead of &mut self. Being explicit seems a little more obvious to me here.

Yea, I see both sides. I'm somewhat motivated by how [rustls does it](https://docs.rs/rustls/0.19.0/rustls/trait.StoresServerSessions.html). It also means someone could chose whether `Mutex` or `RwLock` makes more sense (maybe their workflow expects to rarely receive new cookies, but send some in most requests).

pfernie commented:
I do agree it feels a bit awkward to be exposing a particular locking primitive in the `reqwest` API. My thought on the other side (cmp. to say `rust-tls`) is that `cookie_store::CookieStore` provides many methods handing out `&`s of stored cookies. That's a bit awkward to implement with locking if it's hidden within the implementation, unless we start cloning items.

However, I think it is straightforward for `cookie_store` to simply implement a couple of simple wrappers, `CookieStoreMutex` and `CookieStoreRwLock`, for common usage, where the locking is still explicit for the end-user.

The latest changes remove `Mutex` from the `reqwest` API, and `cookie_store` is updated with these [simple wrappers](https://github.com/pfernie/cookie_store/blob/c2495284380d9301e429916cab2919e821429390/src/lib.rs#L19-L95)

pfernie commented:
This updates to latest `master` (although will probably need another update as `0.11` is readied). The tests have also been updated to handle `reqwest` no longer having a default `CookieStore` available. This allows removing `cookie_store` crate as a dependency under the `cookies` feature. It should, however, add a `dev-dependency` on `cookie_store` in order to have a `CookieStore` implementation for tests. As it stands, since each crate (`reqwest`, `cookie_store`) depends on the other in this scenario, I am unsure of how to specify the dependencies in `Cargo.toml` in order to have it build/pass CI properly, without having the crates published. The tests pass locally when using `path = ` dependency specifications, but specifying `git` dependencies does not work. Alternatively, perhaps a dummy `CookieStore` (naive `HashMap`) implementation should be provided for tests, to avoid the cyclical dependency requirement?

Other than the above, still unaddressed is whether the `CookieStore` trait should use the `HeaderValue` types instead of `&str` as it does currently.

Are there any other questions/issues you would like addressed in this PR?

seanmonstar commented on src/async_impl/client.rs:
```diff
@@ -435,23 +432,22 @@ impl ClientBuilder {
     ///
     /// This requires the optional `cookies` feature to be enabled.
     #[cfg(feature = "cookies")]
-    pub fn cookie_store(mut self, enable: bool) -> ClientBuilder {
```
I completely forgot that this was actually proposing a breaking change. Sigh... Well, for 0.11.x, I guess we can live with `cookie_store(bool)` and `cookie_provider(dyn CookieStore)` (naming hard).

seanmonstar commented:
Merged in #1203

albel727 commented:
Good work! I've been waiting for this for a long time. Now I only wish one didn't have to fully reimplement the Jar in order to support another common use-case: persisting cookies.

pfernie commented:
The [v0.13.0 release of `cookie_store`](https://crates.io/crates/cookie_store/0.13.0) provides a feature, `reqwest_impl` implementing the new trait `reqwest::cookie::CookieStore`. An example can be found [here](https://docs.rs/cookie_store/0.13.0/cookie_store/#feature-reqwest_impl), which illustrates deserializing a `CookieStore` and providing it to a `reqwest::Client` via the new `cookie_provider` API.
